cloudinit/osys/network.py

```diff
@@ -0,0 +1,179 @@
+        """Get an object representing the available interfaces."""
+
+    @abc.abstractmethod
+    def hosts(self):
+        """Get an object representing the available hosts."""
+
+    @abc.abstractmethod
+    def set_hostname(self, hostname):
+        """Change the host name of the instance."""
+
+    @abc.abstractmethod
+    def set_timezone(self, timezone):
+        """Change the current timezone with the given timezone."""
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseRoute(object):
+    """Base class for routes."""
+
+    def __init__(self, destination, gateway, netmask,
+                 interface, metric,
+                 flags=None, refs=None, use=None, expire=None):
+        self.destination = destination
+        self.gateway = gateway
+        self.netmask = netmask
+        self.interface = interface
+        self.metric = metric
+        self.flags = flags
+        self.refs = refs
+        self.use = use
+        self.expire = expire
+
+    @abc.abstractproperty
+    def static(self):
+        """Check if this route is static."""
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseRouteCollection(collections.Sequence):
+    """A collection which encapsulates the routes available in the system.
+
+    Each route item will be an instance of :class:`BaseRoute` and a particular
+    implementation depending on the underlying platform.
+    To retrieve all the routes, it is sufficient to instantiate this class,
+    it will, under the hood, retrieve the routes.
+    The collection offers a couple of methods for adding or deleting a route.
+
+    >>> RouteCollection.add(Route(..., ...))
+    >>> RouteCollection.delete(Route(..., ...))
+    >>> list(RouteCollection())
+    >>> Route(..., ...) in RouteCollection()
+    """
+
+    def __init__(self):
+        self._route_items = self._routes()
+        super(BaseRouteCollection, self).__init__()
+
+    @abc.abstractmethod
+    def _routes(self):
+        """Low level function to retrieve the available routes."""
+
+    @util.abstractclassmethod
+    def add(cls, route):
+        """Add a new route in the underlying OS.
+
+        The function should expect an instance of :class:`BaseRoute`.
+        """
+
+    @util.abstractclassmethod
+    def delete(cls, route):
+        """Delete a route from the underlying OS.
+
+        This function should expect an instance of :class:`BaseRoute`.
+        """
+
+    def __getitem__(self, index):
+        return self._route_items[index]
+
+    def __delitem__(self, index):
+        self.__class__.delete(self._route_items[index])
+        del self._route_items[index]
+
+    def __len__(self):
+        return len(self._route_items)
+
+    def __contains__(self, other):
+        if isinstance(other, BaseRoute):
+            return other in self._route_items
+
+        return any(route.destination == other.destination
+                   for route in self._route_items)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseInterface(object):
```

Comment:
Meh, this is python, people will have to look at the base class code to know this anyway. And docs should tell this also.... Prefixing all the things with 'Base' just adds noise imho.